flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCachePutTest.java

```diff
@@ -582,57 +545,47 @@ public void testPutBufferFailsForJobHa() throws IOException {
      */
     private void testPutBufferFails(@Nullable final JobID jobId, BlobKey.BlobType blobType)
             throws IOException {
-        assumeTrue(!OperatingSystem.isWindows()); // setWritable doesn't work on Windows.
+        // setWritable doesn't work on Windows.
+        assumeThat(OperatingSystem.isWindows()).as("setWritable doesn't work on Windows").isFalse();
 
-        final Configuration config = new Configuration();
-        File tempFileDir = null;
-        try (BlobServer server =
-                        new BlobServer(config, temporaryFolder.newFolder(), new VoidBlobStore());
-                BlobCacheService cache =
-                        new BlobCacheService(
-                                config,
-                                temporaryFolder.newFolder(),
-                                new VoidBlobStore(),
-                                new InetSocketAddress("localhost", server.getPort()))) {
+        Tuple2<BlobServer, BlobCacheService> serverAndCache =
+                TestingBlobUtils.createServerAndCache(tempDir);
 
+        try (BlobServer server = serverAndCache.f0;
+                BlobCacheService cache = serverAndCache.f1) {
             server.start();
 
             // make sure the blob server cannot create any files in its storage dir
-            tempFileDir = server.createTemporaryFilename().getParentFile().getParentFile();
-            assertTrue(tempFileDir.setExecutable(true, false));
-            assertTrue(tempFileDir.setReadable(true, false));
-            assertTrue(tempFileDir.setWritable(false, false));
+            File tempFileDir = server.createTemporaryFilename().getParentFile().getParentFile();
+            assertThat(tempFileDir.setExecutable(true, false)).isTrue();
+            assertThat(tempFileDir.setReadable(true, false)).isTrue();
+            assertThat(tempFileDir.setWritable(false, false)).isTrue();
 
             byte[] data = new byte[2000000];
             rnd.nextBytes(data);
 
-            // upload the file to the server via the cache
-            exception.expect(IOException.class);
-            exception.expectMessage("PUT operation failed: ");
-
-            put(cache, jobId, data, blobType);
-
-        } finally {
-            // set writable again to make sure we can remove the directory
-            if (tempFileDir != null) {
-                //noinspection ResultOfMethodCallIgnored
-                tempFileDir.setWritable(true, false);
+            try {
+                assertThatThrownBy(() -> put(cache, jobId, data, blobType))
+                        .isInstanceOf(IOException.class)
+                        .hasMessageStartingWith("PUT operation failed: ");
+            } finally {
+                assertThat(tempFileDir.setWritable(true, false)).isTrue();
             }
         }
```

Comment:
Actually, that inner `try` is necessary, because the `tempFileDir` write permissions has to be set back before the `close()` method of the `BlobServer` gets called. If the `finally` block is attached to the outer `try`, an `AccessDeniedException` is thrown, cause `close()` is executed before the `finally` block and the dir is not writable yet:
```
java.nio.file.AccessDeniedException: /var/folders/wp/ccy48gw1255bswh9bx9svxjc0000gn/T/junit8517198549335353984/server/incoming
```

Reply:
If so, I have 2 questions:
1. Why does the master branch just have one `try finally`?
2. For master branch, if the outer `try` throw exception, the `tempFileDir.setWritable(true, false);` can be called in the finally `block`. However, the `tempFileDir.setWritable(true, false);` is only called after inner `try` is executed. Is it expected?

Reply:
This is a difference between JUnit4 and JUnit5. The blob server storage folder (where `tempFileDir` points) has no write permissions at the time of `close()` is called on master as well, but JUnit4 ignores that and removes the folder anyways. JUnit5 does not. 

JUnit4 list dir for the server storage dir at the point of delete:
```
/var/folders/wp/ccy48gw1255bswh9bx9svxjc0000gn/T/junit5602027938676775669
% ll
total 0
dr-xr-xr-x  3 fcsaky  staff    96B Aug 23 15:58 junit4915192828668545622
```

JUnit5 list dir for the server storage dir at the point of delete:
```
/var/folders/wp/ccy48gw1255bswh9bx9svxjc0000gn/T/junit2862686155928225667
% ll
total 0
dr-xr-xr-x  3 fcsaky  staff    96B Aug 23 16:08 server
```

Reply:
Thanks @ferenc-csaky  for the detailed clarification, sounds make sense.

We have used the `assertThatThrownBy` to catch exception, so the inner finally can be removed, right? Like this:

```
   private void testPutBufferFails(@Nullable final JobID jobId, BlobKey.BlobType blobType)
            throws IOException {
        assumeThat(OperatingSystem.isWindows()).as("setWritable doesn't work on Windows").isFalse();

        Tuple2<BlobServer, BlobCacheService> serverAndCache =
                TestingBlobUtils.createServerAndCache(tempDir);

        try (BlobServer server = serverAndCache.f0;
                BlobCacheService cache = serverAndCache.f1) {
            server.start();

            // make sure the blob server cannot create any files in its storage dir
            File tempFileDir = server.createTemporaryFilename().getParentFile().getParentFile();
            assertThat(tempFileDir.setExecutable(true, false)).isTrue();
            assertThat(tempFileDir.setReadable(true, false)).isTrue();
            assertThat(tempFileDir.setWritable(false, false)).isTrue();

            byte[] data = new byte[2000000];
            rnd.nextBytes(data);

            assertThatThrownBy(() -> put(cache, jobId, data, blobType))
                    .isInstanceOf(IOException.class)
                    .hasMessageStartingWith("PUT operation failed: ");
            assertThat(tempFileDir.setWritable(true, false)).isTrue();
        }
    }
```

Reply:
Thanks @ferenc-csaky  for the detailed clarification, sounds make sense.

We have used the `assertThatThrownBy` to catch exception, so the inner finally can be removed, right? Like this:

```
   private void testPutBufferFails(@Nullable final JobID jobId, BlobKey.BlobType blobType)
            throws IOException {
        assumeThat(OperatingSystem.isWindows()).as("setWritable doesn't work on Windows").isFalse();

        Tuple2<BlobServer, BlobCacheService> serverAndCache =
                TestingBlobUtils.createServerAndCache(tempDir);

        try (BlobServer server = serverAndCache.f0;
                BlobCacheService cache = serverAndCache.f1) {
            server.start();

            // make sure the blob server cannot create any files in its storage dir
            File tempFileDir = server.createTemporaryFilename().getParentFile().getParentFile();
            assertThat(tempFileDir.setExecutable(true, false)).isTrue();
            assertThat(tempFileDir.setReadable(true, false)).isTrue();
            assertThat(tempFileDir.setWritable(false, false)).isTrue();

            byte[] data = new byte[2000000];
            rnd.nextBytes(data);

            assertThatThrownBy(() -> put(cache, jobId, data, blobType))
                    .isInstanceOf(IOException.class)
                    .hasMessageStartingWith("PUT operation failed: ");
            assertThat(tempFileDir.setWritable(true, false)).isTrue();
        }
    }
```

Please correct me if I'm wrong, thanks a lot~

Reply:
No, you are correct, thanks for the suggestion, fixed!